Add missing infra documentation (#2650)
hashintel/hash: infra/terraform/README.md

By default, the selected region is `us-east-1` and can be configured by editing the TF variables used for applying the TF plan, e.g. the one in [`./hash/prod-usea1.tfvars`](./hash/prod-usea1.tfvars).

# Naming convention

Resources use the following naming convention:

```text
{OWNER}-{PROJECT}-{ENV}-{REGION}-{RESOURCE DESCRIPTION}

OWNER : 'h' for hash
PROJECT : 'hash'
ENV : 'dev' or 'prod'  (maybe others too later)
REGION : AWS region shortened -- 'usea1' etc.
RESOURCE DESCRIPTION : a short description of the resource e.g. 'subnetpub1'
```

> Fields may _not_ include any hyphens. Example names: `h-hash-prod-usea1-vpc`, `h-hash-dev-usea2-apisvc`.

Reviewer: ```suggestion
Fields (e.g. Owner, Project) may _not_ include any hyphens. 

Example _valid_ names: `h-hash-prod-usea1-vpc`, `h-hash-dev-usea2-apisvc`
Example _invalid_ name: `h-hash-prod-usea1-api-svc`
```

> (Inspired by https://stepan.wtf/cloud-naming-convention/)

Reviewer: ```suggestion
(Inspired by [https://stepan.wtf/cloud-naming-convention](https://stepan.wtf/cloud-naming-convention/))
```

The region and environment fields are set by the variables supplied in the `*.tfvars` file. The config automatically shortens AWS region names, for example, `us-east-1` will be converted to `usea1`.

# Deploying

Deployment currently relies on a couple of manual steps - but is to be automated in the future. The container registries need to have images pushed in the correct place for the applications to start and the database has to be migrated manually. The order of executions for deployment:

1.  Deploy infrastructure with terraform
1.  Migrate databases
1.  Build/push docker images

## Deploy infrastructure with terraform

> Secret environment should be provided in Hashicorp Vault. These are expected in a kvv2 path starting with `pipelines/hash/` and ending with the environment name, e.g. `automation/pipelines/hash/prod`. The following secrets are expected:

Reviewer 2: Just a clarification question: With "expected" do you mean these are the required variables for deployment?

Author: Yes, the terraform config assumes/expects the path and the JSON structure to contain the secrets. We can change the wording if you want

Reviewer 2: I think the wording is good. I didn't realize this is actual JSON we expect here 🙂

Author: Oh i see. It is JSON, but the top level fields are treated as key-value pairs in vault. So the nested values are JSON objects, the top level ones are just vault key values :)

> Secret environment should be provided in HashiCorp Vault. These are expected in a kvv2 path starting with `pipelines/hash/` and ending with the environment name, e.g. `automation/pipelines/hash/prod`. The following secrets are expected:

Reviewer: where does `automation` come from here?

Author: That's the engine name, I've tried to make it more clear in 82fe967394043693d34176331d985bef0c8d92ce

```console
$ terraform workspace show
prod
$ terraform apply --var-file prod-usea1.tfvars
..
```

Note that it may be required to disable refreshing state for subsequent applies (because of the Postgres SSH tunnel. Data sources are deferred to the apply phase, usually).

```console

> $ terraform apply --var-file prod-usea1.tfvars -refresh=false

Reviewer 2: ```suggestion
$ terraform apply --var-file prod-usea1.tfvars --refresh=false
```

Author: Huh interesting that both `--` and `-` are valid. I'll apply the change for consistency

Reviewer 2: Oh, I thought it was a typo, didn't know that `-refresh` also works 😄

This will start an SSH tunnel making `localhost:5554` point to the remote RDS instance within the private subnet in AWS.

> To migrate the graph, you must first build the docker container that contains the graph and run it with the graph credentials you should have in the Vault instance.

Reviewer 2: ```suggestion
To migrate `hash-graph`, you must first build the docker container that contains the graph and run it with the graph credentials you should have in the Vault instance.
```

# Troubleshooting and guides

## What to do if the HASH DB needs **Disaster Recovery**

There's a [recovery playbook](./playbooks/db_recovery.md) on what we should do when the DB is on fire, please follow it to recover a snapshot.

## Help! I need access to the private subnet!

You can access the private subnet using the Bastion host. For more information, please see [the corresponding readme](./hash/bastion.md). You need an SSH privatekey to access the bastion, which you can find in `1password` if you have the appropriate access level.

## How do I get the Terraform state from S3?

Install the AWS CLI and follow instructions to add credentials. Credentials are generated in the IAM User security credentials page. `https://us-east-1.console.aws.amazon.com/iam/home#/users/NAME?section=security_credentials` for a user with the name `NAME`.

Once the CLI has been configured, you should have a `.aws` folder in your home directory. Terraform will automatically pick this up and connect to the S3 Terraform State backend.

## How do I log in to AWS ECR container registry?

> If you wish to pull/push container images to ECR, AWS CLI allows you to do so with the following command:
> 
> ```console
> $ aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin ACCOUNTID.dkr.ecr.us-east-1.amazonaws.com
> ..
> ```
> 
> This will log into the given ECR in the `us-east-1` region.

Reviewer: This was missing steps in the process to 'pull/push container images to ECR', suggest we just link to the AWS docs which give them all
```suggestion
If you wish to pull/push container images to ECR manually, you must:
1. Login to the registry
1. Build and tag the image
1. Push the image to ECR
Please see the [AWS docs](https://docs.aws.amazon.com/AmazonECR/latest/userguide/docker-push-ecr-image.html) for instructions on how to do so.
```

where the `TASKID` is the ID of the Fargate service task. Note that this would start an `sh` session directly into your production container, and should be done with caution (if it ever should be done). It's recommended not to run commands against the production environment like this, and we may at some point restrict access to Fargate Exec.

# Troubleshooting and guides

## What to do if the HASH DB needs **Disaster Recovery**

> There's a [playbook](./playbooks/db_recovery.md) on what we should do when the DB is on fire, please follow it to recover a snapshot.

Reviewer 2: ```suggestion
There's a [recovery playbook](./playbooks/db_recovery.md) on what we should do when the DB is on fire, please follow it to recover a snapshot.
```

## Help! I need access to the private subnet!

> You can access the private subnet using the Bastion host. See [this README](./hash/bastion.md) for more information. You need an SSH privatekey to access the bastion, which you can find in `1password` if you have the appropriate access level.

Reviewer 2: ```suggestion
You can access the private subnet using the Bastion host. For more information, please see [the corresponding readme](./hash/bastion.md). You need an SSH privatekey to access the bastion, which you can find in `1password` if you have the appropriate access level.
```

Once the CLI has been configured, you should have a `.aws` folder in your home directory. Terraform will automatically pick this up and connect to the S3 Terraform State backend.

## How do I log in to AWS ECR container registry?

If you wish to pull/push container images to ECR, AWS CLI allows you to do so with the following command:

```console
$ aws ecr get-login-password --region us-east-1 | docker login --username AWS --password-stdin ACCOUNTID.dkr.ecr.us-east-1.amazonaws.com
..
```

This will log into the given ECR in the `us-east-1` region.

## How do I migrate the database after it has been deployed?

Using the Terraform scripts to deploy an instance of RDS will give you an empty DB (with appropriate DBs/users), which will need to be migrated. From the `hashintel/hash` repo you can run the following command assuming you've set up an SSH tunnel (see [this README](./hash/bastion.md)) and have the appropriate environment variables set for the DB connection:

```console
$ docker run --rm \
  --network host \
  -e HASH_GRAPH_PG_USER \
  -e HASH_GRAPH_PG_PASSWORD \
  -e HASH_GRAPH_PG_HOST \
  -e HASH_GRAPH_PG_PORT \

>   -e HASH_GRAPH_PG_DATABASE \

Author: They are, yes. The PG database one isn't though. It has to be read from the output of Terraform. Perhaps I should add explicit steps for setting these values

>   -e HASH_GRAPH_PG_DATABASE \

Reviewer 2: I think this would be helpful, thank you!

>   -e HASH_GRAPH_PG_DATABASE \

Author: Addded in 4066e5e4cf66dd83d7cf7379d7db6ed706cfc11d

>   -e HASH_GRAPH_PG_DATABASE \

Reviewer 2: Great, thank you!

>   -e RUST_LOG \

Reviewer 2: Is it not required to actually pass the environment variables?

- [`bastion`](./modules/bastion/) - a [Bastion host](https://en.wikipedia.org/wiki/Bastion_host) that resides in the public subnet with access to the private subnet. Accessible through SSH.
- [`container_cluster`](./modules/container_cluster/) - is a thin wrapper around the `aws_ecs_cluster` resource with container capacity providers.
- [`container_registry`](./modules/container_registry/) - is a thin wrapper around the `aws_ecr_repository` resource with container evicition policies.
- [`redis`](./modules/redis/) - [external service] a multi Availability Zone Redis cluster with encryption enabled.
- [`tunnel`](./modules/tunnel/) - a custom SSH tunnel using an external data source to allow terraform to connect to services on the private subnet of the VPC.
- [`variables`](./modules/variables/) - contains global variable validation/definitions generally useful for our Terraform infrastructure.
- [`vault_aws_auth`](./modules/vault_aws_auth/) - contains the configuration for authenticating the AWS provider through the Vault AWS authentication backend.

> **HASH specefic modules**:

Reviewer 2: Same as above, if there was a reason just resolve this comment
```suggestion
## HASH specefic modules
```
